parsley-debug/shared/src/main/scala/parsley/debugger/combinator.scala

```diff
@@ -0,0 +1,197 @@
+      * for this debugger.
+      *
+      * It is recommended that you save the debug trees somewhere should you want to re-use the same
+      * debugged parser, or if you want to use the same debugged parser as a child parser on
+      * multiple parts of your main parser, make sure you use [[attachReusable]], as re-use of the
+      * same debugged parser across multiple parent parsers will cause the different parse trees to
+      * incorrectly merge in an undefined manner.
+      *
+      * A small warning: debugging an already debugged parser (via [[attachDebugger]] and friends)
+      * is an undefined behaviour.
+      *
+      * See [[attachWithFrontend]] to automate the tree processing after parsing.
+      *
+      * @note Do not run a parser through this combinator multiple times.
+      *
+      * @param parser The parser to debug.
+      * @param toStringRules If a parser's result matches any of the predicates in this sequence, it
+      *                      will get turned into a string before storing in the debug tree. This is
+      *                      usually for memory-usage optimisation. By default, all function-like
+      *                      objects will be converted into strings, as closures are expensive to store.
+      * @tparam A Output type of original parser.
+      * @return A pair of the finalised tree, and the instrumented parser.
+      */
+    def attachDebugger[A](parser: Parsley[A], toStringRules: Seq[Any => Boolean] = defaultRules): (() => DebugTree, Parsley[A]) = {
+        // XXX: A weak map is needed so that memory leaks will not be caused by flatMap parsers.
+        //      We want a decent amount of initial space to speed up debugging larger parsers slightly.
+        val seen: ParserTracker   = new ParserTracker(new XWeakMap(64)) // scalastyle:ignore magic.number
+        val context: DebugContext = new DebugContext(toStringRules)
+
+        val attached: LazyParsley[A] = injectM[A](parser.internal, seen, context)
+        val resetter: Parsley[Unit]  = fresh(context.reset()).impure
+
+        (() => context.getFinalTree, resetter *> new Parsley(attached))
+    }
+
+    // $COVERAGE-OFF$
+    /** Create a closure that freshly attaches a debugger to a parser every time it is called.
+      * This is used for creating debugged parsers that can be used as children to multiple parent
+      * parsers, as using the same debugged parser as a child to multiple parsers is unsafe.
+      *
+      * See [[attachDebugger]] for more information, or [[attachWithFrontend]] for a version that
+      * also accepts a [[parsley.debugger.frontend.DebugFrontend]].
+      *
+      * @note Do not run a parser through this combinator multiple times.
+      * @return Generator closure for debugged versions of the input parser.
+      */
+    def attachReusable[A](parser: Parsley[A], toStringRules: Seq[Any => Boolean] = defaultRules): () => (() => DebugTree, Parsley[A]) =
```

Comment:
Default params are also a nightmare for the binary evolution: prefer overloaded method signatures instead (you may have to duplicate the documentation a little, see `parsley.debug` for inspiration)

Reply:
Resolved as of [ec940bb](https://github.com/j-mie6/parsley/pull/219/commits/ec940bbd669fa02b28cff5ef64fff33d64b95788)